funsor/distributions.py

```diff
@@ -449,22 +450,81 @@ def eager_mvn(loc, scale_tril, value):
 
 
 # Create a Gaussian from a ground observation.
-@eager.register(MultivariateNormal, Tensor, Tensor, Variable)
-@eager.register(MultivariateNormal, Variable, Tensor, Tensor)
+# TODO refactor this logic into Gaussian.eager_subs() and
+#   here return Gaussian(...scale_tril...)(value=loc-value).
+@eager.register(MultivariateNormal, (Variable, Contraction), Tensor, (Variable, Contraction))
+@eager.register(MultivariateNormal, (Variable, Contraction), Tensor, Tensor)
+@eager.register(MultivariateNormal, Tensor, Tensor, (Variable, Contraction))
 def eager_mvn(loc, scale_tril, value):
-    if isinstance(loc, Variable):
-        loc, value = value, loc
+    assert len(loc.shape) == 1
+    assert len(scale_tril.shape) == 2
+    assert value.output == loc.output
+
+    affine = loc - value
+    if not is_affine(affine):
+        return None  # lazy
+
+    const, coeffs = extract_affine(affine)
+    if not isinstance(const, Tensor):
+        return None  # lazy
+    if not all(isinstance(coeff, Tensor) for coeff, _ in coeffs.values()):
+        return None  # lazy
+
+    # Dovetail to avoid variable name collision in einsum.
+    equations1 = [''.join(c if c in ',->' else chr(ord(c) * 2 - ord('a')) for c in eqn)
+                  for _, eqn in coeffs.values()]
+    equations2 = [''.join(c if c in ',->' else chr(ord(c) * 2 - ord('a') + 1) for c in eqn)
+                  for _, eqn in coeffs.values()]
 
-    dim, = loc.output.shape
-    inputs, (loc, scale_tril) = align_tensors(loc, scale_tril)
-    inputs.update(value.inputs)
-    int_inputs = OrderedDict((k, v) for k, v in inputs.items() if v.dtype != 'real')
+    real_inputs = OrderedDict((k, v) for k, v in affine.inputs.items() if v.dtype == 'real')
+    assert tuple(real_inputs) == tuple(coeffs)
+
+    # Incorporate scale_tril before broadcasting.
+    eye = torch.eye(scale_tril.data.size(-1)).expand(scale_tril.data.shape)
+    prec_sqrt = Tensor(eye.triangular_solve(scale_tril.data, upper=False).solution,
+                       scale_tril.inputs)
+    tensors = []
+    for k, (coeff, eqn) in coeffs.items():
+        shape = (prec_sqrt.shape[-1], real_inputs[k].num_elements)
+        tensors.append((prec_sqrt @ coeff.reshape(shape)).reshape(coeff.shape))
```

Comment:
I thought that affine (const, coef A, ...) is interpreted as const + x.T @ A + ... (based on einsum formula `ab,a->b`). So when the affine is scaled by `scale_tril`, the new coeff will be `A @ prec_sqrt`, where `prec_sqrt = scale_tril^{-1}.T`. If you define `prec_sqrt = scale_tril^{-1}`, then the new coeff will be `A @ prec_sqrt.T`.

However, I think I miss something here because you have resolved shape errors.

Reply:
Thanks! done.